src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,612 @@
+                    } else {
+                        low = guess;
+                    }
+                }
+            } else {
+                if (sim.excess0 > EPSILON) {
+                    if (guess <= 1) break;
+                    high = guess - 1;
+                } else {
+                    if (low == guess) {
+                        if (high <= guess + 1) break;
+                        low = guess;
+                    } else {
+                        low = guess;
+                    }
+                }
+            }
+        }
+
+        if (best.success) {
+            return (true, zeroForOne, best.amountIn, best.amountOut, best.sqrtPriceX96);
+        }
+
+        return (false, zeroForOne, 0, 0, sqrtPriceX96);
+    }
+
+    function _executeSwap(
+        PoolKey memory key,
+        bool zeroForOne,
+        uint256 amountIn
+    ) internal returns (uint160 sqrtPriceX96, uint256 uintOut, uint256 uintIn) {
+        if (amountIn == 0) {
+            (uint160 currentSqrtPrice,,,) = poolManager.getSlot0(key.toId());
+            return (currentSqrtPrice, 0, 0);
+        }
+
+        BalanceDelta delta = poolManager.swap(
+            key,
+            IPoolManager.SwapParams({
+                zeroForOne: zeroForOne,
+                amountSpecified: -int256(amountIn),
+                sqrtPriceLimitX96: zeroForOne ? TickMath.MIN_SQRT_PRICE + 1 : TickMath.MAX_SQRT_PRICE - 1
+            }),
+            new bytes(0)
+        );
+
+        _settleDelta(key, delta);
+        poolManager.settle();
+        _collectDelta(key, delta);
+
+        uintIn = zeroForOne ? _abs(delta.amount0()) : _abs(delta.amount1());
+        uintOut = zeroForOne ? _abs(delta.amount1()) : _abs(delta.amount0());
+
+        (sqrtPriceX96,,,) = poolManager.getSlot0(key.toId());
+        return (sqrtPriceX96, uintOut, uintIn);
+    }
+
+    function _addFullRangeLiquidity(
+        PoolKey memory key,
+        Position storage position,
+        uint256 amount0,
+        uint256 amount1,
+        uint160 sqrtPriceX96
+    ) internal returns (uint256 remaining0, uint256 remaining1) {
+        uint128 liquidityDelta = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtPriceAtTick(position.tickLower),
+            TickMath.getSqrtPriceAtTick(position.tickUpper),
+            amount0,
+            amount1
+        );
+
+        if (liquidityDelta == 0) {
+            return (amount0, amount1);
+        }
+
+        (BalanceDelta balanceDelta,) = poolManager.modifyLiquidity(
+            key,
+            IPoolManager.ModifyLiquidityParams({
+                tickLower: position.tickLower,
+                tickUpper: position.tickUpper,
+                liquidityDelta: int256(uint256(liquidityDelta)),
+                salt: position.salt
+            }),
+            new bytes(0)
+        );
+
+        _settleDelta(key, balanceDelta);
+        poolManager.settle();
+        _collectDelta(key, balanceDelta);
```

Comment:
The `poolManager.settle()` call here should be redundant since we call it already in `_pay`